### DISCUSSION TITLE
Does supabase.auth.signin perform any throttling on requests?

### DISCUSSION THREAD

**Discussion by bill-offertrackr** [supabase/supabase#9605] [Questions]
I'm using passwordless sign in via mobile phone for my application. This requires an OTP code to be sent to the user as part of the process. Since sending the SMS has a cost associated with it, I was wondering if Supabase is doing any throttling on the signin APIs to prevent somebody from calling it millions of times (via my site or directly) and costing me a fortune.

This is the code that ultimately produces the SMS message:

```
supabase.auth.signIn({ phone }, { shouldCreateUser: true });
```

Thanks!

**Comment by princefishthrower** [comment]
This is a massive vulnerability in any Supabase app using the `signIn` method for producing OTPs client side. As far as I know, other than the configurable delay (default of 1 minute) to prevent _the same_ phone number, there is no throttling. However, you can probably see that it becomes trivial for an attacker to choose an arbitrary phone number to avoid this limit.

This should be addressed by also limiting based on the IP address (or even an IP range). Our team is currently looking into this, we will likely start with a Kong based solution: https://docs.konghq.com/hub/kong-inc/rate-limiting/ and go from there.

Regardless, in the future, I think supabase should ship with some sort of configurable rate limiting for this endpoint built-in.

**Comment by ilbertt** [comment]
Together with @princefishthrower, we implemented _basic_ rate limiting on a **self-hosted** Supabase instance by adding these lines to [Kong config](https://github.com/supabase/supabase/blob/master/docker/volumes/api/kong.yml):

```diff
...

services:
  ## Open Auth routes
  - name: auth-v1-open
    ...
  - name: auth-v1-open-authorize
    ...

+  ## Separate OTP Auth route to rate limit it (see plugins below)
+  - name: auth-v1-otp
+    _comment: "GoTrue: /auth/v1/otp -> http://auth:9999/otp"
+    url: http://auth:9999/otp
+    routes:
+      - name: auth-v1-otp
+        strip_path: true
+        paths:
+          - /auth/v1/otp
+    plugins:
+      - name: cors
+      - name: key-auth
+        config:
+          hide_credentials: false
+      - name: acl
+        config:
+          hide_groups_header: true
+          allow:
+            - admin
+            - anon

  ## Secure Auth routes
  - name: auth-v1
    _comment: "GoTrue: /auth/v1/* -> http://auth:9999/*"
    url: http://auth:9999/
    routes:
      - name: auth-v1-all
        strip_path: true
        paths:
          - /auth/v1/
    plugins:
      - name: cors
      - name: key-auth
        config:
          hide_credentials: false
      - name: acl
        config:
          hide_groups_header: true
          allow:
            - admin
            - anon

  ## Secure REST routes
  - name: rest-v1
   ...

+plugins:
+  - name: rate-limiting
+    service: auth-v1-otp
+    config: 
+      second: 1
+      minute: 2
+      hour: 10
+      policy: local
+      limit_by: ip
```

Maybe it makes sense to open a PR for this.

### 02.11.2022
Since Supabase Kong version in self hosted solution [is still 2.x](https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml#L30), make sure to look at the right Kong docs: https://docs.konghq.com/hub/kong-inc/rate-limiting/2.8.x.html.
